### ISSUE TITLE
Error when garbage collecting from non-terminated generator

### ISSUE THREAD

**Issue by davidnewman02** [h5py/h5py#1495]
Versions: 
 * h5py=="any"

If you have an unterminated generator then a `SystemError` is thrown when the garbage collector runs at the end of a module. 
```
import h5py

def yield_groups(filename):
    with h5py.File(filename, 'r') as fh:
        for group in fh:
            yield group

filename = "file_with_10_groups.hdf5"
grp_generator = yield_groups(filename)
next(grp_generator)
```
Raises 
```
Exception ignored in: <generator object yield_groups at 0x7f3fa6690468>
Traceback (most recent call last):
  File "h5py_generator.py", line 6, in yield_groups
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "venv/lib/python3.5/site-packages/h5py/_hl/files.py", line 461, in __exit__
  File "venv/lib/python3.5/site-packages/h5py/_hl/files.py", line 432, in close
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5f.pyx", line 267, in h5py.h5f.get_obj_ids
  File "h5py/h5i.pyx", line 37, in h5py.h5i.wrap_identifier
SystemError: Parent module 'h5py' not loaded, cannot perform relative import
```
This is a pretty minor/niche bug. It seems that the `h5py` module is unloaded by the garbage collector before the file handle is closed, which means that the` __exit__()` on the `File` object cannot be called successfully. It only occurs for unterminated generators, with an open file handle, which are being called at the module level.  

This issue does not occur when if the generator is terminated (either by `grp_generator.close()` or `for i in grp_generator: pass`). It also does not happen if the generator call is in a function e.g.:
```
def print_groups():
    filename = "file_with_10_groups.hdf5"
    grp_generator = yield_groups(filename)
    grp = next(grp_generator)
    print(grp)

print_groups()
```
since the `__exit__()` on the file handler is called when the generator object goes out of scope, before it unloads the h5py library at the end of execution.

**Comment by takluyver** [comment]
Thanks. My initial leaning was to say it's not worth silencing the error, as it can only happen when the interpreter is shutting down anyway. But if you trigger this after writing to a file, is it possible for the file to be corrupted or incompletely written? That might be worth fixing, if so.

**Comment by tacaswell** [comment]
Is the fix to be super aggressive about stashing the modules we need in `close` on the File class?

**Comment by takluyver** [comment]
I think it might need to be where the modules are used, i.e. the `wrap_identifier` function. IIRC, the way things are torn down at interpreter shutdown doesn't follow the usual logic of garbage collection, so keeping references to them somewhere doesn't necessarily keep them alive. It's a long time since I investigated anything like this, though.

https://github.com/h5py/h5py/blob/84d7b7e58177e2c1b63ae37115bb4a88f6a90049/h5py/h5i.pyx#L33-L60

**Comment by takluyver** [comment]
This should be fixed by #1854.
